File: java/core/src/core/mail/client/MasterCacheSerializer.java
/**
 * Author: Timothy Prepscius
 * License: GPLv3 Affero + keep my name in the code!
 */
package mail.client;

import core.callback.Callback;
import mail.client.cache.IndexedCacheSerializer;
import mail.client.cache.Item;
import mail.client.cache.ItemSerializer;
import mail.client.cache.JSON;
import mail.client.model.ModelSerializer;
import mail.client.model.Settings;

public class MasterCacheSerializer implements ItemSerializer
{
	IndexedCacheSerializer indexedCacheSerializer;
	ModelSerializer settingsSerializer;
	
	public MasterCacheSerializer (JSON json)
	{
		indexedCacheSerializer = new IndexedCacheSerializer();
		settingsSerializer = new ModelSerializer(json);
	}
	
	@Override
	public Callback serialize_(Item item) 
	{
		if (item instanceof Settings)
			return settingsSerializer.serialize_(item);
		
		return indexedCacheSerializer.serialize_(item);
	}
	@Override
	public Callback deserialize_(Item item) 
	{
		if (item instanceof Settings)
			return settingsSerializer.deserialize_(item);
		
		return indexedCacheSerializer.deserialize_(item);
	}
	
	
}

File: cpp/core/mailiverse/mail/manager/MasterCacheSerializer.cpp
/**
 * Author: Timothy Prepscius
 * License: BSD + keep my name in the code!
 */

#include "MasterCacheSerializer.h"

using namespace mailiverse::mail::manager;
using namespace mailiverse::mail;
using namespace mailiverse;

MasterCacheSerializer::MasterCacheSerializer (cache::ItemSerializer *storeSerializer)
{
	indexedCacheSerializer = new cache::IndexedCacheSerializer(storeSerializer);
	settingsSerializer = new serializers::ZipSerializer(new serializers::JsonSerializerStandard<model::Settings>());
}

MasterCacheSerializer::~MasterCacheSerializer()
{
	delete indexedCacheSerializer;
	delete settingsSerializer;
}

cache::Value MasterCacheSerializer::serialize(cache::Item *item) 
{
	if (dynamic_cast<model::Settings *>(item))
		return settingsSerializer->serialize(item);
	
	return indexedCacheSerializer->serialize(item);
}

void MasterCacheSerializer::deserialize(cache::Item *item, const cache::Value &value) 
{
	if (dynamic_cast<model::Settings *>(item))
		return settingsSerializer->deserialize(item, value);
	
	return indexedCacheSerializer->deserialize(item, value);
}

